cns/restserver/api.go

```diff
@@ -899,6 +899,109 @@ func (service *HTTPRestService) getNetworkContainerByOrchestratorContext(w http.
+	logger.Request(service.Name, &req, err)
+	if err != nil {
+		logger.Printf("[Azure CNS] getAllNetworkContainers failed with error: %s", err.Error())
+		return
+	}
+
+	var returnCode types.ResponseCode
+	var returnMessage string
+	var ncDetails map[string]cns.ContainerStatus
+	allNcExist := false
+
+	switch r.Method {
+	case "GET":
+		ncDetails, allNcExist = service.getAllNetworkContainersDetails(req.NetworkContainersID)
+
+	default:
+		returnMessage = "[Azure CNS] Error. getAllNetworkContainers did not receive a GET."
+		returnCode = types.InvalidParameter
+	}
+
+	response := cns.GetAllNetworkContainersResponse{
+		ContainerStatus:           ncDetails,
+		AllNetworkContainersExist: allNcExist,
+		Response: cns.Response{
+			ReturnCode: returnCode,
+			Message:    returnMessage,
+		},
+	}
+
+	err = service.Listener.Encode(w, &response)
+	logger.Response(service.Name, response, response.Response.ReturnCode, err)
+}
+
+// replenishNetworkContainer receives and stores the GS for all the NCs of node from DNC
+func (service *HTTPRestService) replenishNetworkContainer(w http.ResponseWriter, r *http.Request) {
+	logger.Printf("[Azure CNS] replenishNetworkContainer")
+
+	var req cns.CreateNetworkContainerRequest
+	err := service.Listener.Decode(w, r, &req)
+	logger.Request(service.Name, req.String(), err)
+	if err != nil {
+		return
+	}
+
+	var returnCode types.ResponseCode
+	var returnMessage string
+
+	switch r.Method {
+	case "POST":
+		if req.NetworkContainerType == cns.WebApps {
+			// Try to get the saved NC state if it exists
+			existing, ok := service.getNetworkContainerDetails(req.NetworkContainerid)
+
+			// Create/update nc only if it doesn't exist, or it exists and the requested version is different from the saved version
+			if !ok || (ok && existing.VMVersion != req.Version) {
+				nc := service.networkContainer
+				if err = nc.Create(req); err != nil {
+					returnMessage = fmt.Sprintf("[Azure CNS] Error. replenishNetworkContainer failed %v", err.Error())
+					returnCode = types.UnexpectedError
+					break
+				}
+			}
+		} else if req.NetworkContainerType == cns.AzureContainerInstance {
+			// Try to get the saved NC state if it exists
+			existing, ok := service.getNetworkContainerDetails(req.NetworkContainerid)
+
+			// Create/update nc only if it doesn't exist, or it exists and the requested version is different from the saved version
+			if ok && existing.VMVersion != req.Version {
+				nc := service.networkContainer
+				netPluginConfig := service.getNetPluginDetails()
+				if err = nc.Update(req, netPluginConfig); err != nil {
+					returnMessage = fmt.Sprintf("[Azure CNS] Error. replenishNetworkContainer failed %v", err.Error())
+					returnCode = types.UnexpectedError
+					break
+				}
+			}
+		}
+
+		returnCode, returnMessage = service.saveNetworkContainerGoalState(req)
+
+	default:
+		returnMessage = "[Azure CNS] Error. replenishNetworkContainer did not receive a POST."
+		returnCode = types.InvalidParameter
+	}
```

Comment:
Same comment about the dataflow. Rethink this and optimize it into a single path.

Reply:
Addressed.